pdf-converter/client.py

```diff
@@ -0,0 +1,427 @@
+                          final=name.with_suffix(f".{final}"))
+
+
+def save_rep(rep, data, size):
+    if rep.initial.write_bytes(data) != size:
+        logging.error("inconsistent initial representation file size")
+        raise RepresentationError
+
+
+async def recv_rep(loop, proc, tmpdir, page):
+    """Receive initial representation from the server
+
+    :param proc: Qrexec process to read STDIN from
+    :param path: File path which will store the initial representation
+    """
+    rep = get_rep(tmpdir, page, "rgb", "png")
+
+    try:
+        dim = await get_img_dim(proc)
+        data = await recv_b(proc, dim.size)
+        await loop.run_in_executor(None, save_rep, rep, data, dim.size)
+    except EOFError as e:
+        raise ReceiveError from e
+    except (DimensionError, ReceiveError, RepresentationError):
+        raise
+
+    return rep, dim
+
+
+async def start_convert(rep, dim):
+    cmd = ["convert", "-size", f"{dim.width}x{dim.height}", "-depth",
+            f"{dim.depth}", f"rgb:{rep.initial}", f"png:{rep.final}"]
+
+    try:
+        proc = await asyncio.create_subprocess_exec(*cmd)
+        await wait_proc(proc)
+    except asyncio.CancelledError:
+        terminate_proc(proc)
+
+    return proc
+
+
+# TODO (?): Save->delete PNGs in loop to avoid storing all PNGs in memory.
+# TODO: Add error handling
+def combine_reps(save_path, reps):
+    with Image.open(reps[0].final) as first:
+        remaining = [Image.open(rep.final) for rep in reps[1:]]
+        first.save(save_path, "PDF", resolution=100, append_images=remaining,
+                   save_all=True)
+
+    for img in remaining:
+        img.close()
+
+
+###############################
+#           Main
+###############################
+
+async def receive(loop, proc, tmpdir):
+    procs = []
+    reps = []
+
+    try:
+        pagenums = await recv_pagenums(loop, proc)
+    except (PageError, ReceiveError):
+        raise
+
+    for page in range(1, pagenums + 1):
+        try:
+            rep, dim = await recv_rep(loop, proc, tmpdir, page)
+        except (DimensionError, ReceiveError, RepresentationError):
+            term_tasks = [terminate_proc(p) for p in procs]
+            await asyncio.gather(*term_tasks)
+            raise
+
+        reps.append(rep)
+        procs.append(await start_convert(rep, dim))
+
+    return procs, reps
+
+
+async def convert(loop, path, procs, reps):
+    for proc, rep in zip(procs, reps):
+        try:
+            await wait_proc(proc)
+        except subprocess.CalledProcessError:
+            logging.error("page conversion failed")
+            raise
+
+        await loop.run_in_executor(None, unlink, rep.initial)
+
+    save_path = path.with_suffix(".trusted.pdf")
+    await loop.run_in_executor(None, combine_reps, save_path, reps)
+
+    return save_path
+
+
+async def sanitize(loop, proc, path):
+    with TemporaryDirectory(prefix=f"qvm-sanitize-{proc.pid}-") as tmpdir:
```

Comment:
When converting multiple files, if a `client-vm` subprocess fails early on, its temporary directory may be cleaned up before the other `client-vm` subprocesses create their directories, leading to a collision. Not likely, but still.

Reply:
TemporaryDirectory will generate unique name (with a given prefix in this case), so even if another one is cleaned up, it isn't a problem.

Reply:
Oh, didn't know that. I guess it's naming scheme is different from `NamedTemporaryFile`? Fixed in ede1a51.

Reply:
It's the same. https://docs.python.org/3/library/tempfile.html:
> Files names used by this module include a string of random characters which allows those files to be securely created in shared temporary directories.

Reply:
What. Didn't you say this earlier though?

> NamedTemporaryFile() specifically create the file in a way that avoid collision. Once it's removed, you can't just reuse the name without extra checks to preserve this property.

I took it as meaning once the file is deleted, `NamedTemporaryFile()` may (in the very unlikely situation) reuse the name.

Not that it matters anyways though since we got to a better solution with `TemporaryDirectory()`.

Reply:
> I took it as meaning once the file is deleted, `NamedTemporaryFile()` may (in the very unlikely situation) reuse the name.

Yes, that's possible, but shouldn't be an issue - after you delete it, you shouldn't use it anymore (including trying to deleting again), so it should be safe to reuse. Anyway, not a problem with `TemporaryDirectory()`.

Reply:
Fixed in ede1a51.